rules/onchain_budget.go

```diff
@@ -0,0 +1,694 @@
+//
+// NOTE: this is part of the Values interface.
+func (o *OnChainBudget) RealToPseudo(db firewalldb.PrivacyMapReader,
+	flags session.PrivacyFlags) (Values, map[string]string, error) {
+
+	return o, nil, nil
+}
+
+// checkFeeRate is a helper function that returns an error if the given fee rate
+// exceeds the maximum allowed by the rule that the enforcer is enforcing.
+func (o *OnChainBudgetEnforcer) checkFeeRate(satPerVByte uint64) error {
+	if satPerVByte > o.MaxSatPerVByte {
+		return fmt.Errorf("max fee rate exceeded")
+	}
+	return nil
+}
+
+// formatReqId creates a unique identifier for the request that can be used to
+// associate the action accross request-response handling.
+func formatReqId(lndConnID string, reqID int64) string {
+	return fmt.Sprintf("%s-%d", lndConnID, reqID)
+}
+
+// handleOpenChannelRequest is a helper function which handles the request for
+// opening a single channel.
+func (o *OnChainBudgetEnforcer) handleOpenChannelRequest(ctx context.Context,
+	r *lnrpc.OpenChannelRequest) (proto.Message, error) {
+
+	if r.FundMax {
+		return nil, fmt.Errorf("max funding not supported")
+	}
+
+	if len(r.Outpoints) > 0 {
+		return nil, fmt.Errorf("outpoints not supported")
+	}
+
+	if r.TargetConf != 0 {
+		return nil, fmt.Errorf("target conf not supported")
+	}
+
+	err := o.checkFeeRate(r.SatPerVbyte)
+	if err != nil {
+		return nil, err
+	}
+
+	// Derive a unique identifier for the request to be able to associate it
+	// with the repsonse.
+	reqID := formatReqId(o.GetLndConnID(), o.GetReqID())
+
+	// Prepend an identifier to later be able to check the action's outcome.
+	r.Memo = fmt.Sprintf("%s:%s", reqID, r.Memo)
+
+	amt := uint64(r.LocalFundingAmount + r.PushSat)
+	err = o.handlePendingPayment(
+		ctx,
+		&onChainAction{
+			Amount:     amt,
+			ActionType: actionTypeOpenChannel,
+		},
+		reqID,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+// handleBatchOpenChannelRequest is a helper function which handles the request
+// for opening a multiple channels.
+func (o *OnChainBudgetEnforcer) handleBatchOpenChannelRequest(
+	ctx context.Context, r *lnrpc.BatchOpenChannelRequest) (proto.Message,
+	error) {
+
+	if r.TargetConf != 0 {
+		return nil, fmt.Errorf("target conf " +
+			"not supported")
+	}
+
+	err := o.checkFeeRate(uint64(r.SatPerVbyte))
+	if err != nil {
+		return nil, err
+	}
+
+	// Derive a unique identifier for the request to be able to associate it
+	// with the repsonse.
+	reqID := formatReqId(o.GetLndConnID(), o.GetReqID())
+
+	// A batch open channel request can contain multiple channels, so we
+	// need to sum up the total amount. This operation is atomic since we
+	// publish a single transaction.
+	var totalAmt int64
+	for _, c := range r.Channels {
+		totalAmt += c.LocalFundingAmount + c.PushSat
+
+		// Prepend an identifier to later be able to check the action's
+		// outcome.
+		c.Memo = fmt.Sprintf("%s:%s", reqID, c.Memo)
```

Comment:
Just leaving a note here, that for batch-channel openings, every channel will have the same Memo. This isn't an issue now, but I think we should consider if this could become an issue for the future if:
1. The memo field could potentially get a unique constraint for channels in the future, effectively breaking this.
2. If lnd is ever updated to enable tracking of channel openings by specifying the memo field, that could become problematic if we don't make it unique.
3. For our intended way of checking on restart if a channel opening was successful or not, I guess this won't become an issue right? I assume that we'd iterate the list of open & closed channels and stop the execution if we find a single channel with this specific memo value, and if so consider all of the pending payment amount to have been confirmed.

Reply:
1. I think it's unlikely that it becomes unique, as that wouldn't be great UX-wise if a user has to account for this.
2. Good point! I guess it would make sense to track based on batched channel opens (as transactions bundle them atomically). The memo field could still be used for that, returning any batch for which a single memo field matched. The memo field isn't that nice for that, I'd rather use the pending channel id, but better to have the recordings here, in case we need to do something with them (we can still ignore them in the future).
3. Right, that would be the approach, we'd fetch all pending items and cross check against the channel API's memos.